Rubberduck.Parsing/VBA/RubberduckParser.cs

```diff
@@ -206,19 +246,71 @@ private void ParseAll()
                 }
             }
 
-            var parseTasks = new Task[components.Count];
-            for (var i = 0; i < components.Count; i++)
+            _projectDeclarations.Clear();
+            _state.ClearBuiltInReferences();
+
+            var parseTasks = new Task[toParse.Count];
+            for (var i = 0; i < toParse.Count; i++)
             {
-                parseTasks[i] = ParseAsync(components[i], CancellationToken.None);
+                var index = i;
+                parseTasks[i] = new Task(() =>
+                {
+                    ParseAsync(toParse[index], CancellationToken.None).Wait();
+
+                    if (_resolverTokenSource.IsCancellationRequested || _central.IsCancellationRequested)
+                    {
+                        return;
+                    }
+
+                    if (_state.Status == ParserState.Error) { return; }
+
+                    var qualifiedName = new QualifiedModuleName(toParse[index]);
+                    Logger.Debug("Module '{0}' {1}", qualifiedName.ComponentName,
+                        _state.IsNewOrModified(qualifiedName) ? "was modified" : "was NOT modified");
+
+                    _state.SetModuleState(toParse[index], ParserState.Resolving);
+                    ResolveDeclarations(qualifiedName.Component,
+                        _state.ParseTrees.Find(s => s.Key == qualifiedName).Value);
+                });
+
+                parseTasks[i].Start();
             }
 
             Task.WaitAll(parseTasks);
+            _state.SetStatusAndFireStateChanged(ParserState.ResolvedDeclarations);
+
+            if (_state.Status < ParserState.Error)
+            {
+                ResolveReferencesAsync();
+            }
+        }
+
+        private Task[] ResolveReferencesAsync()
+        {
+            var finder = new DeclarationFinder(_state.AllDeclarations, _state.AllComments, _state.AllAnnotations);
+            var passes = new List<ICompilationPass>
+                {
+                    // This pass has to come first because the type binding resolution depends on it.
+                    new ProjectReferencePass(finder),
+                    new TypeHierarchyPass(finder, new VBAExpressionParser()),
+                    new TypeAnnotationPass(finder, new VBAExpressionParser())
+                };
+            passes.ForEach(p => p.Execute());
+
+            var tasks = new Task[_state.ParseTrees.Count];
 
-            if (_state.Status != ParserState.Error)
```

Comment:
You need a `Debug.Assert` call if you're going to remove that condition: identifier reference resolution should not be allowed to run under any circumstances when the parser state is `Error`.